### PR TITLE
Allow `protected` methods to be called between types in the same namespace

### PR DESCRIPTION
**Pull request by asterite**
Check the specs in this PR to understand the behaviour.

The motivation behind this is that it frequently happens that you have a set of types under a single namespace and you need to invoke some methods between them, but you don't want them to be public. I believe in Ruby you either make them public but undocumented, or you use `send`. Neither solution is elegant.

Some concrete examples in the standard library:
1. [Channel](https://github.com/manastech/crystal/blob/90e8ecd173d8febccf6b1c0ab35b8b140d3b429a/src/concurrent/channel.cr) defines [some operations](https://github.com/manastech/crystal/blob/90e8ecd173d8febccf6b1c0ab35b8b140d3b429a/src/concurrent/channel.cr#L112) which have a reference to a channel and need to invoke [some methods](https://github.com/manastech/crystal/blob/90e8ecd173d8febccf6b1c0ab35b8b140d3b429a/src/concurrent/channel.cr#L125). The only way to make that work is to make those methods public but undocumented (ugly), but it makes sense for the internal class to be able to see protected methods of the namespace it's in, and viceversa.
2. [Hash](https://github.com/manastech/crystal/blob/90e8ecd173d8febccf6b1c0ab35b8b140d3b429a/src/hash.cr) has [entries](https://github.com/manastech/crystal/blob/90e8ecd173d8febccf6b1c0ab35b8b140d3b429a/src/hash.cr#L777), and it makes sense for Hash to be able to invoke protected methods on Entry, because it knows Entry.
3. The runtime will have a Scheduler, Fiber and Event types that need some coordination and co-operation. We'll probably put all of those under a `Crystal::Runtime` namespace, and this will allow these types to cooperate by using protected methods.

Another motivation is that when you define a shard, you generally define new types under a single namespace, and it makes sense for these types to know each other and sometimes invoke some methods between them, even though those aren't public.

Basically, `protected` will now mean "protected by a namespace". Classes outside the namespace won't be able to invoke these methods.

An alternative solution would be to introduce a new keyword, but I think just `protected` is enough, instead of having one for "inside the same type hierarchy" and "inside the same namespace".

We'd like some feedback on this, but we think it's a sound change.

### CODE REVIEW

**Comment by jhass** [comment]
Do I follow right that

```cr
module A
  class Foo
    protected def bar
    end
  end
end

class Bar < Foo
  def initialize
    bar
  end
end

Bar.new
```

will no longer work then? Or does `protected` get both of Java's visibilities, protected and package protected?

**Comment by asterite** [comment]
@jhass It will still work, it gets both meanings. In fact, if you extend a class under a namespace and put it in another namespace, you still have access to the old namespace's protected methods. The new meaning is pretty relaxed.

**Comment by ozra** [comment]
Maybe it's better to ditch the traditional names, if they're gonna work differently (which I'm all for).

`internal` sounds good to me for 'this very type only'.
Some far out ideas (to breed better ideas):
`family` (just made up now - not a very good one) for 'this and all subtypes'
`community` - haha, ok, now the examples are really deteriorating, but you catch the gist - for 'all types within module'.

Or why the hell not - reversed order from above:
`inpub`/`inall`, `inmod`, `intype`, `inself`
or
`forpub`/`forall`, `formod`, `fortype`, `forself`

**Comment by asterite** [comment]
@ozra `protected` already works in a different way than Ruby, because you can invoke class methods from instance methods and viceversa. So changing its semantic again is OK.

**Comment by ozra** [comment]
alright.
